system_metrics_collector/test/system_metrics_collector/test_linux_memory_measurement.cpp

```diff
@@ -143,37 +227,210 @@ class LinuxMemoryMeasurementTestFixture : public ::testing::Test
+  std::promise<bool> empty_promise;
+  std::shared_future<bool> dummy_future = empty_promise.get_future();
+  rclcpp::executors::SingleThreadedExecutor ex;
+  ex.add_node(test_measure_linux_memory);
+  ex.add_node(test_receive_measurements);
+
+  //
+  // spin the node with it started
+  //
+  bool start_success = test_measure_linux_memory->start();
+  ASSERT_TRUE(start_success);
+  ASSERT_TRUE(test_measure_linux_memory->isStarted());
+  ex.spin_until_future_complete(dummy_future, TEST_LENGTH);
+  EXPECT_EQ(3, test_receive_measurements->getNumReceived());
+  // expectation is:
+  // 50 ms: SAMPLES[0] is collected
+  // 80 ms: statistics derived from SAMPLES[0] is published. statistics are cleared
+  // 100 ms: SAMPLES[1] is collected
+  // 150 ms: SAMPLES[2] is collected
+  // 160 ms: statistics derived from SAMPLES[1 & 2] is published. statistics are cleared
+  // 200 ms: SAMPLES[3] is collected
+  // 240 ms: statistics derived from SAMPLES[3] is published. statistics are cleared
+  // 250 ms: SAMPLES[4] is collected. last getStatisticsResults() is of SAMPLES[4]
+  StatisticData data = test_measure_linux_memory->getStatisticsResults();
+  EXPECT_EQ(1, data.sample_count);
+
+  //
+  // spin the node with it stopped
+  //
+  bool stop_success = test_measure_linux_memory->stop();
+  ASSERT_TRUE(stop_success);
+  ASSERT_FALSE(test_measure_linux_memory->isStarted());
+  ex.spin_until_future_complete(dummy_future, TEST_LENGTH);
+  EXPECT_EQ(3, test_receive_measurements->getNumReceived());
+  // expectation is:
+  // upon calling stop, samples are cleared, so getStatisticsResults() would be NaNs
+  // no MetricsMessages are published
+  data = test_measure_linux_memory->getStatisticsResults();
+  EXPECT_TRUE(std::isnan(data.average));
+  EXPECT_TRUE(std::isnan(data.min));
+  EXPECT_TRUE(std::isnan(data.max));
+  EXPECT_TRUE(std::isnan(data.standard_deviation));
+  EXPECT_EQ(0, data.sample_count);
+
+  //
+  // spin the node with it restarted
+  //
+  start_success = test_measure_linux_memory->start();
+  ASSERT_TRUE(start_success);
+  ASSERT_TRUE(test_measure_linux_memory->isStarted());
+  ex.spin_until_future_complete(dummy_future, TEST_LENGTH);
+  EXPECT_EQ(6, test_receive_measurements->getNumReceived());
+  // expectation is:
+  // 50 ms: SAMPLES[5] is collected
+  // 80 ms: statistics derived from SAMPLES[5] is published. statistics are cleared
+  // 100 ms: SAMPLES[6] is collected
+  // 150 ms: SAMPLES[7] is collected
+  // 160 ms: statistics derived from SAMPLES[6 & 7] is published. statistics are cleared
+  // 200 ms: SAMPLES[8] is collected
+  // 240 ms: statistics derived from SAMPLES[8] is published. statistics are cleared
+  // 250 ms: SAMPLES[9] is collected. last getStatisticsResults() is of SAMPLES[9]
+  data = test_measure_linux_memory->getStatisticsResults();
+  EXPECT_EQ(1, data.sample_count);
+
+  //
+  // stop the node to finish off the test
+  //
+  stop_success = test_measure_linux_memory->stop();
```

Comment:
never stop in test cases, do it in Fixture TearDown function

Reply:
Done.